### PR TITLE
vim-patch:8.1.2297: the ex_vimgrep() function is too long

### PR DESCRIPTION
**Pull request by zeertzjq**
#### vim-patch:8.1.2297: the ex_vimgrep() function is too long

Problem:    The ex_vimgrep() function is too long.
Solution:   Split it in three parts. (Yegappan Lakshmanan, closes vim/vim#5211)
https://github.com/vim/vim/commit/d6a98a3a9768568b668f91a53267b36f86b84466

Including a missing change to ex_vimgrep() from patch 8.0.1831.

### CODE REVIEW

File: src/nvim/quickfix.c [hunk 1]
```diff
@@ -5328,118 +5337,85 @@ static bool existing_swapfile(const buf_T *buf)
...
 
   if (eap->addr_count > 0) {
-    tomatch = eap->line2;
+    args->tomatch = eap->line2;
   } else {
-    tomatch = MAXLNUM;
+    args->tomatch = MAXLNUM;
   }
 
-  // Get the search pattern: either white-separated or enclosed in //
```
Comment (seandewar) [inline src/nvim/quickfix.c:5377]: Keep this comment

File: src/nvim/quickfix.c [hunk 2]
```diff
@@ -5522,36 +5504,87 @@ void ex_vimgrep(exarg_T *eap)
...
+  if (au_name != NULL && apply_autocmds(EVENT_QUICKFIXCMDPRE, au_name,
+                                        curbuf->b_fname, true, curbuf)) {
+    if (aborting()) {
+      return;
+    }
+  }
+
+  win_T *wp = NULL;
+  qf_info_T *qi = qf_cmd_get_or_alloc_stack(eap, &wp);
+  char_u *target_dir = NULL;
```
Comment (seandewar) [inline src/nvim/quickfix.c:5531]: This can be moved further down to where `status` is

File: src/nvim/quickfix.c [hunk 3]
```diff
@@ -5522,36 +5504,87 @@ void ex_vimgrep(exarg_T *eap)
...
+  buf_T *first_match_buf = NULL;
+  int status = vgr_process_files(wp, qi, &args, &redraw_for_dummy, &first_match_buf, &target_dir);
+
+  if (status != OK) {
+    FreeWild(args.fcount, args.fnames);
+    decr_quickfix_busy();
+    goto theend;
+  }
+
+  FreeWild(args.fcount, args.fnames);
```
Comment (seandewar) [inline src/nvim/quickfix.c:5556]: Doesn't really matter, but Vim could've just called FreeWild before the `if` 😄

File: src/nvim/quickfix.c (hunk 2 above)
Comment (zeertzjq) [reply src/nvim/quickfix.c:5531]: It needs to be initialized before the first `goto`.

Comment (seandewar) [reply src/nvim/quickfix.c:5531]: Ah, sorry; missed that.

File: src/nvim/quickfix.c (hunk 1 above)
Comment (zeertzjq) [reply src/nvim/quickfix.c:5377]: Oops. Thanks!

File: src/nvim/quickfix.c (hunk 3 above)
Comment (zeertzjq) [reply src/nvim/quickfix.c:5557]: I think I'll keep this.

Comment (zeertzjq) [reply src/nvim/quickfix.c:5557]: I think I'll keep this.

**Review by seandewar — APPROVED** [review]
LGTM!

File: src/nvim/quickfix.c (hunk 3 above)
Comment (seandewar) [reply src/nvim/quickfix.c:5557]: No problem, was just an observation :+1:

**Pull request merged**